### DISCUSSION TITLE
RabbitMQ shutdown on noproc reason

### DISCUSSION THREAD

**Discussion by aravindchidambaram** [rabbitmq/rabbitmq-server#3902] [Other]
I am using the RabbitMQ Server 3.9.5
and Erlang 24

```
2021-11-13 07:16:44.894000+05:30 [info] <0.23125.6> connection <0.23125.6> ([::1]:52350 -> [::1]:5672): user 'guest' authenticated and granted access to vhost '/'
2021-11-13 07:16:44.895000+05:30 [info] <0.23125.6> closing AMQP connection <0.23125.6> ([::1]:52350 -> [::1]:5672, vhost: '/', user: 'guest')
2021-11-13 07:16:44.896000+05:30 [error] <0.23131.6>     supervisor: {<0.23131.6>,rabbit_channel_sup}
2021-11-13 07:16:44.896000+05:30 [error] <0.23131.6>     errorContext: shutdown_error
2021-11-13 07:16:44.896000+05:30 [error] <0.23131.6>     reason: noproc
2021-11-13 07:16:44.896000+05:30 [error] <0.23131.6>     offender: [{pid,<0.23134.6>},
2021-11-13 07:16:44.896000+05:30 [error] <0.23131.6>                {id,channel},
2021-11-13 07:16:44.896000+05:30 [error] <0.23131.6>                {mfargs,
2021-11-13 07:16:44.896000+05:30 [error] <0.23131.6>                    {rabbit_channel,start_link,
2021-11-13 07:16:44.896000+05:30 [error] <0.23131.6>                        [1,<0.23125.6>,<0.23132.6>,<0.23125.6>,
2021-11-13 07:16:44.896000+05:30 [error] <0.23131.6>                         <<"[::1]:52350 -> [::1]:5672">>,
2021-11-13 07:16:44.896000+05:30 [error] <0.23131.6>                         rabbit_framing_amqp_0_9_1,
2021-11-13 07:16:44.896000+05:30 [error] <0.23131.6>                         {user,<<"guest">>,
2021-11-13 07:16:44.896000+05:30 [error] <0.23131.6>                             [administrator],
2021-11-13 07:16:44.896000+05:30 [error] <0.23131.6>                             [{rabbit_auth_backend_internal,none}]},
2021-11-13 07:16:44.896000+05:30 [error] <0.23131.6>                         <<"/">>,
2021-11-13 07:16:44.896000+05:30 [error] <0.23131.6>                         [{<<"publisher_confirms">>,bool,true},
2021-11-13 07:16:44.896000+05:30 [error] <0.23131.6>                          {<<"exchange_exchange_bindings">>,bool,true},
2021-11-13 07:16:44.896000+05:30 [error] <0.23131.6>                          {<<"basic.nack">>,bool,true},
2021-11-13 07:16:44.896000+05:30 [error] <0.23131.6>                          {<<"consumer_cancel_notify">>,bool,true},
2021-11-13 07:16:44.896000+05:30 [error] <0.23131.6>                          {<<"connection.blocked">>,bool,true},
2021-11-13 07:16:44.896000+05:30 [error] <0.23131.6>                          {<<"authentication_failure_close">>,bool,true}],
2021-11-13 07:16:44.896000+05:30 [error] <0.23131.6>                         <0.23126.6>,<0.23133.6>]}},
2021-11-13 07:16:44.896000+05:30 [error] <0.23131.6>                {restart_type,intrinsic},
2021-11-13 07:16:44.896000+05:30 [error] <0.23131.6>                {shutdown,70000},
2021-11-13 07:16:44.896000+05:30 [error] <0.23131.6>                {child_type,worker}]
2021-11-13 07:16:48.397000+05:30 [info] <0.23144.6> accepting AMQP connection <0.23144.6> ([::1]:52351 -> [::1]:5672)
```

**Comment by lhoguin** [comment]
You can ignore it. This will be gone once `supervisor2` has been brought up to date with `supervisor`.

**Reply by JamesTGrant** [reply]
I see the same issues in my RabbitMQ cluster when under load, I'm interested in your comment regarding supervisor2 vs supervisor - how can I find out more about this?

**Reply by michaelklishin** [reply]
`supervisor2` and `supervisor` are implementation details for RabbitMQ users. Unless you are familiar with Erlang, assume that a very old module in RabbitMQ has to be updated to match certain expectations of modern Erlang versions.

**Reply by JamesTGrant** [reply]
Thank you for replying. When might this be - a current version or in planning/development? Is there a thread I can watch/follow? I see similar errors in my logs. I suspect that these events correspond to messages not reaching clients and non-deterministic behaviour results. A sample of my logs, do you think this is intrisic or can I tune RabbitMQ to avoid this (there are plenty of memory, CPU and file descriptor resources availalble): (search for '[erro]')
[sadlogs.txt](https://github.com/rabbitmq/rabbitmq-server/files/8958572/sadlogs.txt)

**Reply by michaelklishin** [reply]
There is nothing that can be tuned.

**Reply by lukebakken** [reply]
Please fix the high connection churn!

```
2022-06-21 13:39:34.230370+00:00 [info] <0.14247.2> accepting AMQP connection <0.14247.2> (10.4.48.3:57108 -> 10.4.0.15:5672)
2022-06-21 13:39:34.250466+00:00 [info] <0.14247.2> connection <0.14247.2> (10.4.48.3:57108 -> 10.4.0.15:5672): user 'guest' authenticated and granted access to vhost '/'
2022-06-21 13:39:34.709914+00:00 [info] <0.14217.2> closing AMQP connection <0.14217.2> (10.4.0.42:42848 -> 10.4.0.15:5672, vhost: '/', user: 'guest')
```

**Reply by lukebakken** [reply]
I also see that your client application or perhaps load balancer (if you're using one) is unexpectedly closing connections:

```
2022-06-21 14:42:59.600694+00:00 [warn] <0.12972.2> client unexpectedly closed TCP connection
2022-06-21 14:42:59.601238+00:00 [erro] <0.12983.2>     supervisor: {<0.12983.2>,rabbit_channel_sup}
2022-06-21 14:42:59.601238+00:00 [erro] <0.12983.2>     errorContext: shutdown_error
2022-06-21 14:42:59.601238+00:00 [erro] <0.12983.2>     reason: noproc
```

**Reply by lukebakken** [reply]
Based on these logs your #1 priority is to fix your client applications to use long-lived connections and channels.

**Reply by JamesTGrant** [reply]
Thank you very much for your replies. We have several RabbitMQ client libraries, some better than others. I suspect that one of them implements heartbeats incorrectly - as soon as the heartbeat is complete, the object is deleted causing the socket to be destroyed, or something like that - which I'll get addressed.
For your interest, the behaviour in the logs occurs very reproducably on a scale up, scale down test where we deploy n pods and the application in the pods make connections to RabbitMQ at about(ish) the same time.
I estimate it it will take me a until mid week next week to get these changes made (am out for a few days). Will report back with results when changes are made. Is this GitHub discussion still a good place to capture this?

**Reply by lukebakken** [reply]
> Is this GitHub discussion still a good place to capture this?

Sure! GH discussions aren't great but they work well enough.

**Reply by lukebakken** [reply]
> the behaviour in the logs occurs very reproducably on a scale up, scale down test where we deploy n pods and the application in the pods make connections to RabbitMQ at about(ish) the same time.

If the log message is always `client unexpectedly closed TCP connection` then that is what it is - the _client_ is closing the connection, not RabbitMQ.

**Reply by michaelklishin** [reply]
By the way, that message is [mentioned in the doc section on connection lifecycle events](https://rabbitmq.com/logging.html#logged-events).

**Reply by JamesTGrant** [reply]
Hi all, I have been doing some testing and have results I'd like to share with you. Please find attached a log output. In this test, I have removed connection churn.
The test is performed using a single RabbitMQ server node (to remove complexity), a number of RabbitMQ clients are connected and remain connected throughout the test.

I have a script which:
1 A RabbitMQ client creates two connections with the RabbitMQ server
2 sleep 30sec
3 The RabbitMQ client application is exited such that the TCP socket layer is closed, causing the client end to send a TCP fin to the RabbitMQ server, which sees the TCP connections have gone and so closes the corresponding connections. I'm aware this isn't the most graceful way to close connections, but it's what happens 'in real life' when a k8s pod is deleted (ie; scaled down or restarted) (unless the containers in the pod have elegant SIGTERM handling, which most seem not to)
4 sleep 30sec
5 back to 1

In the attached logs, I start a sequence of this test at 2022-06-30 09:13:06.727889+00:00
You will see that the logs report an erro at 2022-06-30 09:42:26.081381+00:00 and 2022-06-30 09:46:27.699323+00:00
Logs are 'rabbitmqProcesscrashOnClosingConnection.txt'

Earlier in the logs I do a little 'spot testing' and you can see some processes crashes at 2022-06-30 08:09:27.727405+00:00 and 2022-06-30 08:19:27.720494+00:00 and 2022-06-30 08:29:27.721794+00:00

For this test, I am using RabbitMQ server version 3.9.5 in the bitnami container bitnami/rabbitmq:3.9.5-debian-10-r12 I have also attached logs showing the RabbitMQ server startup, perhaps this is helpful.
[rabbitMQServerStartLog.txt](https://github.com/rabbitmq/rabbitmq-server/files/9018903/rabbitMQServerStartLog.txt)

The node on which the RabbitMQ server is running is 32core, and is mostly idle - with lots of availalble memory (over 90GB Free), lots of availalble file desciptors per process etc.

[rabbitmqProcesscrashOnClosingConnection.txt](https://github.com/rabbitmq/rabbitmq-server/files/9018744/rabbitmqProcesscrashOnClosingConnection.txt)

I will repeat my test with a RabbitMQ client which creates and closes connections to the RabbitMQ server at the RabbitMQ client application layer (rather then TCP layer triggered connection closure) with regular periodicy and will report back.

Is there more I can do to?

**Reply by JamesTGrant** [reply]
Here's a nice BASH and Python harness which can be used to show the problem relably. 
Running the first /tmp/pyfun - which has the release(), never causes RabbitMQ server process crash.
Running the 2nd /tmp/pyfun - which causes the RabbitMQ server to do the disconnection due to the client TCP socket closure regularly causes RabbitMQ server process crashes. Within an hour I'd expect to see two or three.

``` 
#!/bin/bash

#This is BASH

#this does not show the problem, RabbitMQ client disconnects nicely
cat << EOF > /tmp/pyfun
from kombu import Connection
import time
server_url = "amqp://guest:guest@rabbitmq-ha:5672/?heartbeat=30"
c1 = Connection(server_url)
c2 = Connection(server_url)
c1.connect()
c2.connect()
time.sleep(30)
c1.release()
c2.release()
exit()
EOF

#this does show the problem, RabbitMQ client connection closed at TCP layer, RabbitMQ closes coresponding connection
cat << EOF > /tmp/pyfun
from kombu import Connection
import time
server_url = "amqp://guest:guest@rabbitmq-ha:5672/?heartbeat=30"
c1 = Connection(server_url)
c2 = Connection(server_url)
c1.connect()
c2.connect()
time.sleep(30)
exit()
EOF

while true; do
/bin/python3 /tmp/pyfun
sleep 30
done
```

**Reply by JamesTGrant** [reply]
@michaelklishin @lukebakken I hope you are both well. I'm happy that I've provided a pathological test harness (see code above) to show the issue in any current RabbitMQ cluster of one or more nodes. I think the problem is real and can lead to messages getting lost from the system - and I think we agree the problem is within RabbitMQ/Erlang. Whilst I can mitigate the likelyhood of provoking this bug in my systems by improving client behaviour, I can not prevent this occuring as cluster degridation lifecycle events will cause the RabbitMQ/Server to exercise this (sometimes errant) functionality. Are you happy to acknowledge this is a bug and, if so is there a way I can track it? Kindest regards and thank you.

**Reply by michaelklishin** [reply]
I'm afraid I do not acknowledge that this is a bug in either RabbitMQ or Erlang.

This is a behavior of your client which abruptly terminates. RabbitMQ detects this condition
and closes its end of the connection. This can happen in any realistic
distributed system. I am not sure what your expectations are.

Because abrupt TCP connection closure can be problematic for publishers, there is a protocol
feature and a connection recovery procedure that is provided by some clients or can be
re-created manually in others. This is documented in [Publishers guide](https://www.rabbitmq.com/publishers.html#data-safety) as well as the guide on [consumer acknowledgements and publisher confirms](https://www.rabbitmq.com/confirms.html). Your example uses neither of those features,
so data loss is possible. There is nothing RabbitMQ or Erlang cannot do about client
publishers not getting through. RabbitMQ can and does acknowledge publishes it has received
back to the client.

I am not certain that Kombu supports publisher confirms. A few years ago it did not support
a lot of data safety and operational features. It is a client we would highly recommend
against in favor of [Pika](https://github.com/pika/pika/) which does support publisher confirms.

**Reply by JamesTGrant** [reply]
hi @michaelklishin - thank you for the reply. It is very helpful. My example above is the most simple way to cause the erro log to output the messages that concern me - it is simply a harness to create connections, and have the connections be unexpectedly lost. I am aware that there is no acknowledge/confirm for this example - it's just a simple example to provide the most simple harness to provoke the erro condition.
My question is: The erro logs which sometimes (not every time) appear as a result of the unexpected connection loss. These do not happen on every unexpected TCP connection closure. Some of these logs appear to be process crashes, it's these occasional crashes which I'm suggesting are a bug. (https://github.com/rabbitmq/rabbitmq-server/files/9018744/rabbitmqProcesscrashOnClosingConnection.txt) for example:

```
2022-06-30 08:09:27.726750+00:00 [warn] <0.27844.233> client unexpectedly closed TCP connection
2022-06-30 08:09:27.727405+00:00 [erro] <0.27837.233>     supervisor: {<0.27837.233>,rabbit_channel_sup_sup}
2022-06-30 08:09:27.727405+00:00 [erro] <0.27837.233>     errorContext: shutdown_error
2022-06-30 08:09:27.727405+00:00 [erro] <0.27837.233>     reason: shutdown
2022-06-30 08:09:27.727405+00:00 [erro] <0.27837.233>     offender: [{nb_children,1},
2022-06-30 08:09:27.727405+00:00 [erro] <0.27837.233>                {id,channel_sup},
2022-06-30 08:09:27.727405+00:00 [erro] <0.27837.233>                {mfargs,{rabbit_channel_sup,start_link,[]}},
2022-06-30 08:09:27.727405+00:00 [erro] <0.27837.233>                {restart_type,temporary},
2022-06-30 08:09:27.727405+00:00 [erro] <0.27837.233>                {shutdown,infinity},
2022-06-30 08:09:27.727405+00:00 [erro] <0.27837.233>                {child_type,supervisor}]

2022-06-30 08:19:27.719969+00:00 [warn] <0.28670.233> client unexpectedly closed TCP connection
2022-06-30 08:19:27.720494+00:00 [erro] <0.28938.233>     supervisor: {<0.28938.233>,rabbit_channel_sup}
2022-06-30 08:19:27.720494+00:00 [erro] <0.28938.233>     errorContext: shutdown_error
2022-06-30 08:19:27.720494+00:00 [erro] <0.28938.233>     reason: noproc
2022-06-30 08:19:27.720494+00:00 [erro] <0.28938.233>     offender: [{pid,<0.28954.233>},
2022-06-30 08:19:27.720494+00:00 [erro] <0.28938.233>                {id,channel},
2022-06-30 08:19:27.720494+00:00 [erro] <0.28938.233>                {mfargs,
2022-06-30 08:19:27.720494+00:00 [erro] <0.28938.233>                    {rabbit_channel,start_link,
2022-06-30 08:19:27.720494+00:00 [erro] <0.28938.233>                        [1,<0.28670.233>,<0.28940.233>,<0.28670.233>,
2022-06-30 08:19:27.720494+00:00 [erro] <0.28938.233>                         <<"10.1.0.15:29267 -> 10.4.0.15:5672">>,
2022-06-30 08:19:27.720494+00:00 [erro] <0.28938.233>                         rabbit_framing_amqp_0_9_1,
2022-06-30 08:19:27.720494+00:00 [erro] <0.28938.233>                         {user,<<"guest">>,
2022-06-30 08:19:27.720494+00:00 [erro] <0.28938.233>                             [administrator],
2022-06-30 08:19:27.720494+00:00 [erro] <0.28938.233>                             [{rabbit_auth_backend_internal,none}]},
2022-06-30 08:19:27.720494+00:00 [erro] <0.28938.233>                         <<"/">>,
2022-06-30 08:19:27.720494+00:00 [erro] <0.28938.233>                         [{<<"authentication_failure_close">>,bool,true},
2022-06-30 08:19:27.720494+00:00 [erro] <0.28938.233>                          {<<"basic.nack">>,bool,true},
2022-06-30 08:19:27.720494+00:00 [erro] <0.28938.233>                          {<<"connection.blocked">>,bool,false},
2022-06-30 08:19:27.720494+00:00 [erro] <0.28938.233>                          {<<"consumer_cancel_notify">>,bool,true},
2022-06-30 08:19:27.720494+00:00 [erro] <0.28938.233>                          {<<"publisher_confirms">>,bool,true}],
2022-06-30 08:19:27.720494+00:00 [erro] <0.28938.233>                         <0.28929.233>,<0.28943.233>]}},
2022-06-30 08:19:27.720494+00:00 [erro] <0.28938.233>                {restart_type,intrinsic},
2022-06-30 08:19:27.720494+00:00 [erro] <0.28938.233>                {shutdown,70000},
2022-06-30 08:19:27.720494+00:00 [erro] <0.28938.233>                {child_type,worker}]
```

**Reply by michaelklishin** [reply]
Yes, getting rid of this error report (logged by the runtime) makes sense. We have a couple of ideas, it's not a matter of removing a log line in our own code, unfortunately. We will have to port some Erlang 24 changes to one of our bundled modules.

**Comment by aravindchidambaram** [comment]
But the port 15672 is not listening when I run the rabbitmq-diagnostics status command and so for same, I couldn't able to access my management dashboard at localhost:15672

**Comment by michaelklishin** [comment]
I will convert this issue to a GitHub discussion. Currently GitHub will [automatically close and lock the issue](https://github.com/github/feedback/discussions/3197) even though your question will be transferred and responded to elsewhere. This is to let you know that we do not intend to ignore this but this is how the current GitHub conversion mechanism makes it seem for the users :(

**Comment by michaelklishin** [comment]
@aravindchidambaram the exception you have shared happened when a connection was closed. There is also evidence of a connection being accepted at the end of the log.

We have no evidence of this exception being connected to any `rabbitmq-diagnostics` output at all but [**port 15672 is not used for AMQP 0-9-1/AMQP 1.0 connections**](От меня немного отстал саппорт, сейчас опишу, как прошла туристическая часть моего дня) which you are looking at in the log. It is used by the HTTP API and management UI that must be enabled via a plugin.

**Comment by jipock-morgan** [comment]
I see this hasn't been officially answered yet. I was experiencing this very error message yesterday on our development server and after a change in a custom library, the error stopped happening. The problem was frequent, yet intermittent. When it occurred, RabbitMQ's log had similar entries. 

**TL;DR:** 
**Make sure you don't have some sort of other functions / processes that are closing the connection on your end when they shouldn't be.** 

Versions: 
RabbitMQ version: 3.11.2  
Erlang 25.0.2
RabbitMQ.Client (for .NET 6 project): 6.4.0

We have a more developer-friendly wrapper library over the RabbitMQ.Client, which is built to add logging and be a little more friendly to developers when consuming / publishing to RabbitMQ.  **The issue for me was that this wrapper library was clearing connection / channel objects prior to creating a RabbitMQ consumer.**  Ordinarily, this should be a problem, since it the first time it is run, there shouldn't be any connection / channels present yet. But, in an application using this library, it was making both channel / connections in another thread. The wrapper library does not use static functions or singletons, so I don't know why this would happen (I haven't checked the RabbitMQ.Client source code yet), but taking out this connection clearing routine fixed the issue. There also are not lingering connections either, so it turned out to be entirely unnecessary. 

<details><summary>Error log</summary>

```
2022-11-28 16:05:03.275000-05:00 [warning] <0.11737.35> closing AMQP connection <0.11737.35> (10.242.193.206:64722 -> 10.142.231.1:5671 - VM-DEVMACHINE - Virtual Host: cache - Exchange: cache.resources, vhost: 'cache', user: 'cache_user'):
2022-11-28 16:05:03.275000-05:00 [warning] <0.11737.35> client unexpectedly closed TCP connection
2022-11-28 16:05:03.275000-05:00 [error] <0.11743.35>     supervisor: {<0.11743.35>,rabbit_channel_sup}
2022-11-28 16:05:03.275000-05:00 [error] <0.11743.35>     errorContext: shutdown_error
2022-11-28 16:05:03.275000-05:00 [error] <0.11743.35>     reason: noproc
2022-11-28 16:05:03.275000-05:00 [error] <0.11743.35>     offender: [{pid,<0.11746.35>},
2022-11-28 16:05:03.275000-05:00 [error] <0.11743.35>                {id,channel},
2022-11-28 16:05:03.275000-05:00 [error] <0.11743.35>                {mfargs,
2022-11-28 16:05:03.275000-05:00 [error] <0.11743.35>                    {rabbit_channel,start_link,
2022-11-28 16:05:03.275000-05:00 [error] <0.11743.35>                        [1,<0.11737.35>,<0.11740.35>,<0.11737.35>,
2022-11-28 16:05:03.275000-05:00 [error] <0.11743.35>                         <<"10.242.193.1:64722 -> 10.142.231.1:5671">>,
2022-11-28 16:05:03.275000-05:00 [error] <0.11743.35>                         rabbit_framing_amqp_0_9_1,
2022-11-28 16:05:03.275000-05:00 [error] <0.11743.35>                         {user,<<"cache_user">>,[],
2022-11-28 16:05:03.275000-05:00 [error] <0.11743.35>                             [{rabbit_auth_backend_internal,none}]},
2022-11-28 16:05:03.275000-05:00 [error] <0.11743.35>                         <<"cache">>,
2022-11-28 16:05:03.275000-05:00 [error] <0.11743.35>                         [{<<"publisher_confirms">>,bool,true},
2022-11-28 16:05:03.275000-05:00 [error] <0.11743.35>                          {<<"exchange_exchange_bindings">>,bool,true},
2022-11-28 16:05:03.275000-05:00 [error] <0.11743.35>                          {<<"basic.nack">>,bool,true},
2022-11-28 16:05:03.275000-05:00 [error] <0.11743.35>                          {<<"consumer_cancel_notify">>,bool,true},
2022-11-28 16:05:03.275000-05:00 [error] <0.11743.35>                          {<<"connection.blocked">>,bool,true},
2022-11-28 16:05:03.275000-05:00 [error] <0.11743.35>                          {<<"authentication_failure_close">>,bool,true}],
2022-11-28 16:05:03.275000-05:00 [error] <0.11743.35>                         <0.11738.35>,<0.11745.35>]}},
2022-11-28 16:05:03.275000-05:00 [error] <0.11743.35>                {restart_type,intrinsic},
2022-11-28 16:05:03.275000-05:00 [error] <0.11743.35>                {shutdown,70000},
2022-11-28 16:05:03.275000-05:00 [error] <0.11743.35>                {child_type,worker}]
```
</details>

**Reply by lukebakken** [reply]
This is the relevant message:

```
client unexpectedly closed TCP connection
```

Which, is exactly what your client did 😄 It must not be closing / disposing of RabbitMQ.Client objects correctly.

Also, check out how I re-formatted the error log you provided.

**Reply by jipock-morgan** [reply]
Yep - Thank you for the tip.  I'll have to remember the details / summary tags.

Perhaps your reply and my TL;DR might save someone some needless debugging in the future.
